framework/midi/midievent.h

```diff
@@ -0,0 +1,956 @@
+        }
+        case ChannelVoice20: {
+            uint32_t mask = value << 24;
+            m_data[1] &= 0x00FFFFFF;
+            m_data[1] |= mask;
+            break;
+        }
+        default: assert(false);
+        }
+    }
+
+    uint16_t bank() const
+    {
+        assertOpcode({ ProgramChange, RegisteredController, AssignableController, RelativeRegisteredController,
+                       RelativeAssignableController });
+        assertMessageType({ ChannelVoice20 });
+        if (opcode() == ProgramChange) {
+            return ((m_data[1] & 0x7F00) >> 1) | (m_data[1] & 0x7F);
+        }
+        return (m_data[0] >> 8) & 0x7F;
+    }
+
+    void setBank(uint16_t bank)
+    {
+        assertOpcode({ ProgramChange, RegisteredController, AssignableController, RelativeRegisteredController,
+                       RelativeAssignableController });
+        assertMessageType({ ChannelVoice20 });
+        if (opcode() == ProgramChange) {
+            m_data[0] |= 0x01; //set BankValid bit
+            uint32_t mask = (bank & 0x3F80) << 8 | (bank & 0x7F);
+            m_data[1] &= 0xFFFF0000;
+            m_data[1] |= mask;
+            return;
+        }
+        assert(bank < 128);
+        uint32_t mask = bank << 8;
+        m_data[0] &= 0xFFFF00FF;
+        m_data[0] |= mask;
+    }
+
+    bool isBankValid() const
+    {
+        assertOpcode({ ProgramChange });
+        assertMessageType({ ChannelVoice20 });
+        return m_data[0] & 0x01;
+    }
+
+    AttributeType attributeType() const
+    {
+        assertOpcode({ Opcode::NoteOn, Opcode::NoteOff });
+        assertMessageType({ ChannelVoice20 });
+        return static_cast<AttributeType>(m_data[0] & 0xFF);
+    }
+
+    void setAttributeType(AttributeType type)
+    {
+        assertOpcode({ Opcode::NoteOn, Opcode::NoteOff });
+        assertMessageType({ ChannelVoice20 });
+        m_data[0] &= 0xFFFFFF00;
+        m_data[0] |= type;
+    }
+
+    void setAttributeType(uint8_t type) { setAttributeType(static_cast<AttributeType>(type)); }
+
+    uint16_t attribute() const
+    {
+        assertOpcode({ NoteOn, NoteOff });
+        if (messageType() == ChannelVoice20) {
+            return m_data[1] & 0xFFFF;
+        }
+        return 0x00;
+    }
+
+    void setAttribute(uint16_t value)
+    {
+        assertOpcode({ NoteOn, NoteOff });
+        assertMessageType({ ChannelVoice20 });
+        m_data[1] &= 0xFFFF0000;
+        m_data[1] |= value;
+    }
+
+    void setPerNoteDetach(bool value)
+    {
+        assertOpcode({ PerNoteManagement });
+        assertMessageType({ ChannelVoice20 });
+        m_data[0] &= 0xFFFFFFFD;
+        m_data[0] |= value;
+    }
+
+    void setPerNoteReset(bool value)
+    {
+        assertOpcode({ PerNoteManagement });
+        assertMessageType({ ChannelVoice20 });
+        m_data[0] &= 0xFFFFFFFE;
+        m_data[0] |= value;
+    }
+
+    //!convert ChannelVoice from MIDI2.0 to MIDI1.0
+    std::list<Event> toMIDI1_0() const
```

Comment:
fromMIDI**10**Package, but toMIDI**1_0** 

I think it's better to do it uniformly

Reply:
Fixed